url433bec1e56_sarobe_DifficultyInvestigation_tgz-pJ8-fr_inria_optimization_cmaes_examples_CMAExample1J8/src/fr/inria/optimization/cmaes/CMAEvolutionStrategy.java

```diff
@@ -0,0 +1,2704 @@
+        }
+        if ((s = properties.getProperty("cCov")) != null) {
+            sp.setCcov(Double.parseDouble(options.getFirstToken(s)));
+        }
+        
+    }
+//    private void infoVerbose(String s) {
+//        println(" CMA-ES info: " + s);
+//    }
+    
+    private void warning(String s) {
+        println(" CMA-ES warning: " + s);
+    }
+    private void error(String s) { // somehow a relict from the C history of this code
+        println(" CMA-ES error: " + s);
+        //e.printStackTrace();            // output goes to System.err
+        //e.printStackTrace(System.out);  // send trace to stdout
+
+        throw new CMAException(" CMA-ES error: " + s);
+        //      System.exit(-1); 
+    }
+    
+    /** some simple math utilities */
+    class MyMath { // implements java.io.Serializable {
+        int itest;
+        
+        double square(double d) {
+            return d*d;
+        }
+        double prod(double []ar) {
+            double res = 1.0;
+            for(int i = 0; i < ar.length; ++i)
+                res *= ar[i];
+            return res;
+        }
+        
+        public double median(double ar[]) {
+            // need a copy of ar
+            double [] ar2 = new double[ar.length];
+            for (int i = 0; i < ar.length; ++i)
+                ar2[i] = ar[i];
+            Arrays.sort(ar2);
+            if (ar2.length % 2 == 0)
+                return (ar2[ar.length/2] + ar2[ar.length/2-1]) / 2.;
+            else    
+                return ar2[ar.length/2];
+        }
+        
+        /** @return Maximum value of 1-D double array */
+        public double max(double ar[]) {
+            int i;
+            double m;
+            m = ar[0];
+            for (i = 1; i < ar.length; ++i) {
+                if (m < ar[i])
+                    m = ar[i];
+            }
+            return m;
+        }
+
+        /** sqrt(a^2 + b^2) without under/overflow. **/
+        public double hypot(double a, double b) {
+            double r  = 0;
+            if (Math.abs(a) > Math.abs(b)) {
+               r = b/a;
+               r = Math.abs(a)*Math.sqrt(1+r*r);
+            } else if (b != 0) {
+               r = a/b;
+               r = Math.abs(b)*Math.sqrt(1+r*r);
+            }
+            return r;
+         }
+        /** @return index of minium value of 1-D double array */
+        public int minidx(double ar[]) {
+            return minidx(ar, ar.length-1);
+        }
+        
+        /** @return index of minium value of 1-D double 
+         *   array between index 0 and maxidx 
+         * @param ar double[] 
+         * @param maxidx last index to be considered */
+        public int minidx(double[] ar, int maxidx) {
+            int i, idx;
+            idx = 0;
+            for (i = 1; i < maxidx; ++i) {
+                if (ar[idx] > ar[i])
+                    idx = i;
+            }
+            return idx;
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 938:943, 963:968.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.